SpringProject/src/main/java/com/gradproject2019/users/service/UserServiceImpl.java

```diff
@@ -15,22 +18,62 @@
 
     private final UserRepository userRepository;
 
+    public static final String PASSWORD_PATTERN = "((?=.*[a-z])(?=.*[0-9])(?=.*[!?\\#@^&£$*+;:~])(?=.*[A-Z]).{8,16})";
+    public static final String EMAIL_PATTERN = "^[a-zA-Z0-9\\.\\!\\#\\$\\%\\&\\'\\*\\+\\-\\/\\=\\?\\^\\_\\`]+@[a-zA-Z0-9]+\\.[\\.A-Za-z]{1,10}";
+    public static final String USERNAME_PATTERN = "^[a-zA-Z0-9]*$";
+
     public UserServiceImpl(UserRepository userRepository) {
         this.userRepository = userRepository;
     }
 
     @Override
     public UserResponseDto saveUser(UserRequestDto userRequestDto) {
-        User user = from(userRequestDto);
-        String password = user.getPassword();
+        String password = userRequestDto.getPassword();
+        String username = userRequestDto.getUsername();
+        String email = userRequestDto.getEmail();
+
+        usernameExists(username);
+        emailExists(email);
+
         checkPassword(password);
-        user.setPassword(PasswordUtils.hash(password));
+        checkUsername(username);
+        checkEmail(email);
```

Comment:
Seems you're checking the email/username multiple times - these could be extracted out into their own methods (and both verification methods could be extracted out into it's own method, maybe something like `verifyUser`?

Reply:
Done!